doc/reference/trd105-time.md

```diff
@@ -0,0 +1,455 @@
+
+```rust
+pub trait OverflowClient {
+  fn overflow(&self);
+}
+
+pub trait Counter<'a>: Time {
+  fn start(&self) -> Result<(), ErrorCode>;
+  fn stop(&self) -> Result<(), ErrorCode>;
+  fn reset(&self) -> Result<(), ErrorCode>;
+  fn is_running(&self) -> bool;
+  fn set_overflow_client(&'a self, &'a dyn OverflowClient);
+}
+```
+
+The `OverflowClient` trait is separated from the `AlarmClient` trait
+because there are cases when software simply wants a free-running
+counter to keep track of time, but does not need triggers at a
+particular time. For hardware that has a limited number of
+compare registers, allocating one of them when the compare itself
+isn't needed would be wasteful.
+
+Note that Tock's concurrency model means interrupt bottom halves
+can be delayed until the current bottom half (or syscall
+invocation) completes. This means that an overflow callback can
+seem to occur *after* an overflow. For example, suppose there is
+an 8-bit counter. The following execution is possible:
+
+  1. Client code calls Time::now, which returns 250.
+  1. An overflow happens, marking an interrupt as pending but the bottom half doesn't execute yet.
+  1. Client code calls Time::now, which returns 12.
+  1. The main event loop runs, invoking the bottom half.
+  1. The Counter calls OverflowClient::overflow, notifying the client of the overflow.
+
+A `Counter` implementation MUST NOT provide a `Frequency` of a higher
+resolution than an underlying hardware counter. For example, if the
+underlying hardware counter has a frequency of 32 kHz, then a `Counter`
+cannot say it has a frequency of 1MHz by multiplying the underlying
+counter by 32. A `Counter` implementation MAY provide a `Frequency` of
+a lower resolution (e.g., by stripping bits).
+
+The `reset` method of `Counter` resets the counter to 0.
+
+4 `Alarm` and `AlarmClient` traits
+===============================
+
+Instances of the `Alarm` trait track an incrementing clock and can
+trigger callbacks when the clock reaches a specific value as well as
+when it overflows. The trait is derived from `Time` trait and
+therefore has associated `Time::Frequency` and `Ticks` types.
+
+The `AlarmClient` trait handles callbacks from an instance of `Alarm`.
+The trait derives from `OverflowClient` and adds an additional callback
+denoting that the time specified to the `Alarm` has been reached.
+
+`Alarm` and `Timer` (presented below) differ in their level of
+abstraction. An `Alarm` presents the abstraction of receiving a
+callback when a point in time is reached or on an overflow. In
+contrast, `Timer` allows one to request callbacks at some interval in
+the future, either once or periodically. `Alarm` requests a callback
+at an absolute moment while `Timer` requests a callback at a point
+relative to now.
+
+```rust
+pub trait AlarmClient {
+  fn alarm(&self);
+}
+
+pub trait Alarm: Time {
+  fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks);
```

Comment:
If we want to add a `max_dt` method or specify how `Alarm` and/or `Timer` respond to `dt` values that are too large, this would be the time!

Reply:
I think of the dt at more of an issue with the virtual alarm implementation. Without the virtual alarm layer in between, then the actually hardware driver has no problems with a max dt call. I lean toward leaving the TRD unchanged and making the virtual alarm handle the case properly.